### DISCUSSION TITLE
[Denormalizer] How to enable checking if incoming data fills every field of DTO

### DISCUSSION THREAD

**Discussion by fluffycondor** [symfony/symfony#43660] [Q&A] [answered]
Hello all!
I tried to search my problem in the docs, duckduckgo and google but didn't find anything, maybe I'm lacking suitable keywords.
Looks like by default denormalizer silently allows to not fill any field if absent. I want to turn off this behavior and throw an exception if that happens.
This is my code, simplified:
````php
class Service
{
    private DenormalizerInterface $denormalizer;

    public function __construct(DenormalizerInterface $denormalizer)
    {
        $this->denormalizer = $denormalizer;
    }

    public function __invoke(mixed $data): FooDTO
    {
        return $this->denormalizer->denormalize($data, FooDTO::class, context: [
            AbstractNormalizer::ALLOW_EXTRA_ATTRIBUTES => false,
        ]);
    }
}

class FooDTO
{
    public string $bar;
    public string $baz;
}
````
If I pass an array like ['bar' => 'test'] to it, the denormalizer silently returns to me an instance of FooDTO with uninitialized $baz.
How to change this behavior in order to throw an exception instead?
Any help is appreciated.

**Comment by mbabker** [comment, marked as answer]
You should use the validator to check the data after it has been deserialized into your DTO, not rely on the serializer to do this type of validation.

**Reply by fluffycondor** [reply]
Thank you for the answer!
I consider calling a validator after DTO hydration as a fragile practice: you can easily forget to do that, and also you have a DTO with invalid state.
I found solution with PHP 8 constructor property promotion.
This example:
````php
class FooDTO
{
    public string $bar;
    public string $baz;
}
````
becomes
````php
class FooDTO
{
    public function __construct(
        public string $bar,
        public string $baz,
    ) {}
}
````
And it works like a charm. You'll get `Cannot create an instance of "FooDTO" from serialized data because its constructor requires parameter "baz" to be present.` if e.g. `baz` key is absent.

**Reply by lyrixx** [reply]
Validator is the way to go IMHO (you may want to validate is an url property holds an url, etc)


But indeed you can use CPP, and with SF 5.4, error reporting will be even better 👍🏼
